File: 18.js
// double_hash_table.js
class DoubleHashTable {
  static EMPTY = 0;
  static OCC = 1;
  static DEL = 2;

  constructor(m) {
    this.m = m;
    this.keys = new Array(m).fill(null);
    this.state = new Array(m).fill(DoubleHashTable.EMPTY);
  }

  h1(k){ let v = k % this.m; if (v < 0) v += this.m; return v; }
  h2(k){ return 1 + (Math.abs(k) % (this.m - 1)); }

  search(k){
    const a = this.h1(k), b = this.h2(k);
    for(let i=0;i<this.m;i++){
      const idx = (a + i*b) % this.m;
      if (this.state[idx] === DoubleHashTable.EMPTY) return false;
      if (this.state[idx] === DoubleHashTable.OCC && this.keys[idx] === k) return true;
    }
    return false;
  }

  insert(k){
    if (this.search(k)) return;
    const a = this.h1(k), b = this.h2(k);
    let firstDel = -1;
    for(let i=0;i<this.m;i++){
      const idx = (a + i*b) % this.m;
      if (this.state[idx] === DoubleHashTable.EMPTY) {
        const put = firstDel !== -1 ? firstDel : idx;
        this.keys[put] = k; this.state[put] = DoubleHashTable.OCC; return;
      }
      if (this.state[idx] === DoubleHashTable.DEL && firstDel === -1) firstDel = idx;
    }
    if (firstDel !== -1) { this.keys[firstDel] = k; this.state[firstDel] = DoubleHashTable.OCC; }
  }

  delete(k){
    const a = this.h1(k), b = this.h2(k);
    for(let i=0;i<this.m;i++){
      const idx = (a + i*b) % this.m;
      if (this.state[idx] === DoubleHashTable.EMPTY) return;
      if (this.state[idx] === DoubleHashTable.OCC && this.keys[idx] === k) {
        this.state[idx] = DoubleHashTable.DEL; this.keys[idx] = null; return;
      }
    }
  }

  dump(){
    return this.state.map((s,i) => s===DoubleHashTable.OCC ? String(this.keys[i]) : "EMPTY").join(" ");
  }
}

// Example driver for Problem 1 reading from stdin:
if (require.main === module) {
  const fs = require('fs');
  const data = fs.readFileSync(0,'utf8').trim().split(/\s+/);
  if (!data.length) process.exit(0);
  let idx = 0;
  const m = parseInt(data[idx++],10), n = parseInt(data[idx++],10);
  const ht = new DoubleHashTable(m);
  for (let i=0;i<n;i++){
    const cmd = data[idx++];
    if (cmd === "INSERT") { ht.insert(parseInt(data[idx++],10)); }
    else if (cmd === "DELETE") { ht.delete(parseInt(data[idx++],10)); }
    else if (cmd === "SEARCH") { console.log(ht.search(parseInt(data[idx++],10)) ? "YES" : "NO"); }
    else if (cmd === "PRINT") { console.log(ht.dump()); }
  }
}

File: 18.py
# double_hash_table.py
class DoubleHashTable:
    EMPTY = 0
    OCC = 1
    DEL = 2

    def __init__(self, m):
        self.m = m
        self.keys = [None] * m
        self.state = [DoubleHashTable.EMPTY] * m

    def _h1(self, k):
        return k % self.m

    def _h2(self, k):
        return 1 + (abs(k) % (self.m - 1))

    def search(self, k):
        h1 = self._h1(k)
        h2 = self._h2(k)
        for i in range(self.m):
            idx = (h1 + i * h2) % self.m
            if self.state[idx] == DoubleHashTable.EMPTY:
                return False
            if self.state[idx] == DoubleHashTable.OCC and self.keys[idx] == k:
                return True
        return False

    def insert(self, k):
        if self.search(k):
            return
        h1 = self._h1(k)
        h2 = self._h2(k)
        first_del = -1
        for i in range(self.m):
            idx = (h1 + i * h2) % self.m
            if self.state[idx] == DoubleHashTable.EMPTY:
                if first_del != -1:
                    idx = first_del
                self.keys[idx] = k
                self.state[idx] = DoubleHashTable.OCC
                return
            if self.state[idx] == DoubleHashTable.DEL and first_del == -1:
                first_del = idx
        # table full or rehash needed (not implemented)
        # If first_del available, insert there (already handled)

    def delete(self, k):
        h1 = self._h1(k)
        h2 = self._h2(k)
        for i in range(self.m):
            idx = (h1 + i * h2) % self.m
            if self.state[idx] == DoubleHashTable.EMPTY:
                return
            if self.state[idx] == DoubleHashTable.OCC and self.keys[idx] == k:
                self.state[idx] = DoubleHashTable.DEL
                self.keys[idx] = None
                return

    def dump(self):
        out = []
        for i in range(self.m):
            if self.state[i] == DoubleHashTable.OCC:
                out.append(str(self.keys[i]))
            else:
                out.append("EMPTY")
        return " ".join(out)


def run_problem1():
    import sys
    data = sys.stdin.read().strip().split()
    if not data: return
    it = iter(data)
    m = int(next(it)); n = int(next(it))
    ht = DoubleHashTable(m)
    for _ in range(n):
        cmd = next(it)
        if cmd == "INSERT":
            x = int(next(it)); ht.insert(x)
        elif cmd == "DELETE":
            x = int(next(it)); ht.delete(x)
        elif cmd == "SEARCH":
            x = int(next(it)); print("YES" if ht.search(x) else "NO")
        elif cmd == "PRINT":
            print(ht.dump())


if __name__ == "__main__":
    run_problem1()
